// @twilio/replay-event-emitter/dist/replay-event-emitter.js
/*
@license
Copyright (c) 2021 Twilio Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('./node_modules/tslib/tslib.es6.js');
var _polyfillNode_events = require('./_virtual/_polyfill-node.events.js');

/* eslint-disable @typescript-eslint/no-explicit-any */
var ReplayEventEmitter = /** @class */ (function (_super) {
    tslib_es6.__extends(ReplayEventEmitter, _super);
    function ReplayEventEmitter() {
        var _this = _super.call(this) || this;
        _this.eventHistory = new Map();
        return _this;
    }
    ReplayEventEmitter.prototype.on = function (event, listener) {
        return _super.prototype.on.call(this, event, listener);
    };
    ReplayEventEmitter.prototype.once = function (event, listener) {
        return _super.prototype.once.call(this, event, listener);
    };
    ReplayEventEmitter.prototype.off = function (event, listener) {
        return _super.prototype.off.call(this, event, listener);
    };
    ReplayEventEmitter.prototype.emit = function (event) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.eventHistory.set(event, args);
        return _super.prototype.emit.apply(this, tslib_es6.__spreadArray([event], args));
    };
    ReplayEventEmitter.prototype.addListener = function (event, listener) {
        return _super.prototype.addListener.call(this, event, listener);
    };
    ReplayEventEmitter.prototype.removeListener = function (event, listener) {
        return _super.prototype.removeListener.call(this, event, listener);
    };
    ReplayEventEmitter.prototype.addListenerWithReplay = function (event, listener) {
        var historyEntry = this.eventHistory.get(event);
        if (historyEntry !== undefined) {
            listener.apply(void 0, historyEntry);
        }
        return this.addListener(event, listener);
    };
    ReplayEventEmitter.prototype.onWithReplay = function (event, listener) {
        return this.addListenerWithReplay(event, listener);
    };
    ReplayEventEmitter.prototype.onceWithReplay = function (event, listener) {
        var historyEntry = this.eventHistory.get(event);
        if (historyEntry !== undefined) {
            listener.apply(void 0, historyEntry);
            return this;
        }
        return _super.prototype.once.call(this, event, listener);
    };
    return ReplayEventEmitter;
}(_polyfillNode_events["default"]));

exports.ReplayEventEmitter = ReplayEventEmitter;
