// @twilio/mcs-client/dist/logger.js
/*
@license
Copyright (c) 2018, Twilio, Inc.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

*/
'use strict';

var global =
  typeof global !== "undefined"
    ? global
    : typeof self !== "undefined"
    ? self
    : typeof window !== "undefined"
    ? window
    : {};

Object.defineProperty(exports, '__esModule', { value: true });

var log = require('loglevel');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var log__namespace = /*#__PURE__*/_interopNamespace(log);

function prepareLine(prefix, args) {
    return [`${new Date().toISOString()} MCS Client ${prefix}:`].concat(Array.from(args));
}
class Logger {
    constructor(prefix) {
        this.prefix = "";
        this.prefix =
            prefix !== null && prefix !== undefined && prefix.length > 0
                ? prefix + " "
                : "";
    }
    static scope(prefix) {
        return new Logger(prefix);
    }
    setLevel(level) {
        log__namespace.setLevel(level);
    }
    static setLevel(level) {
        log__namespace.setLevel(level);
    }
    /* eslint-disable @typescript-eslint/explicit-module-boundary-types */
    trace(...args) {
        log__namespace.trace.apply(null, prepareLine(this.prefix + "T", args));
    }
    debug(...args) {
        log__namespace.debug.apply(null, prepareLine(this.prefix + "D", args));
    }
    info(...args) {
        log__namespace.info.apply(null, prepareLine(this.prefix + "I", args));
    }
    warn(...args) {
        log__namespace.warn.apply(null, prepareLine(this.prefix + "W", args));
    }
    error(...args) {
        log__namespace.error.apply(null, prepareLine(this.prefix + "E", args));
    }
    static trace(...args) {
        log__namespace.trace.apply(null, prepareLine("T", args));
    }
    static debug(...args) {
        log__namespace.debug.apply(null, prepareLine("D", args));
    }
    static info(...args) {
        log__namespace.info.apply(null, prepareLine("I", args));
    }
    static warn(...args) {
        log__namespace.warn.apply(null, prepareLine("W", args));
    }
    static error(...args) {
        log__namespace.error.apply(null, prepareLine("E", args));
    }
}

exports.Logger = Logger;
