// twilio-chat/dist/user.js
/*
@license
The following license applies to all parts of this software except as
documented below.

    Copyright (c) 2019, Twilio, inc.
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

      1. Redistributions of source code must retain the above copyright
         notice, this list of conditions and the following disclaimer.

      2. Redistributions in binary form must reproduce the above copyright
         notice, this list of conditions and the following disclaimer in
         the documentation and/or other materials provided with the
         distribution.

      3. Neither the name of Twilio nor the names of its contributors may
         be used to endorse or promote products derived from this software
         without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

This software includes javascript-state-machine under the following license.

    Copyright (c) 2012, 2013, 2014, 2015, Jake Gordon and contributors

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE

This software includes loglevel under the following license.

    Copyright (c) 2013 Tim Perry

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal in the Software without
    restriction, including without limitation the rights to use,
    copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following
    conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
    OTHER DEALINGS IN THE SOFTWARE.

This software includes q under the following license.

    Copyright 2009–2014 Kristopher Michael Kowal. All rights reserved.
    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to
    deal in the Software without restriction, including without limitation the
    rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
    sell copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
    IN THE SOFTWARE.

This software includes platform.js under the following license.

    Copyright 2014 Benjamin Tan <https://d10.github.io/>
    Copyright 2011-2015 John-David Dalton <http://allyoucanleet.com/>

    Permission is hereby granted, free of charge, to any person obtaining
    a copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
    LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
    OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
    WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('./node_modules/tslib/tslib.es6.js');
var logger = require('./logger.js');
var index = require('./util/index.js');
var declarativeTypeValidator = require('@twilio/declarative-type-validator');
var isEqual = require('lodash.isequal');
var replayEventEmitter = require('@twilio/replay-event-emitter');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var isEqual__default = /*#__PURE__*/_interopDefaultLegacy(isEqual);

const log = logger.Logger.scope('User');
/**
 * Extended user information.
 * Note that `isOnline` and `isNotifiable` properties are eligible
 * for use only if the reachability function is enabled.
 * You may check if it is enabled by reading the value of {@link Client.reachabilityEnabled}.
 */
class User extends replayEventEmitter.ReplayEventEmitter {
    /**
     * @internal
     */
    constructor(identity, entityName, configuration, services) {
        super();
        this.promiseToFetch = null;
        /**
         * Fired when the properties or the reachability status of the message have been updated.
         *
         * Parameters:
         * 1. object `data` - info object provided with the event. It has the following properties:
         *     * {@link User} `user` - the user in question
         *     * {@link UserUpdateReason}[] `updateReasons` - array of reasons for the update
         * @event
         */
        this.updated = 'updated';
        /**
         * Fired when the client has subscribed to the user.
         *
         * Parameters:
         * 1. {@link User} `user` - the user in question
         * @event
         */
        this.userSubscribed = 'userSubscribed';
        /**
         * Fired when the client has unsubscribed from the user.
         *
         * Parameters:
         * 1. {@link User} `user` - the user in question
         * @event
         */
        this.userUnsubscribed = 'userUnsubscribed';
        this.services = services;
        this.subscribed = 'initializing';
        this.setMaxListeners(0);
        this.state = {
            identity,
            entityName,
            friendlyName: null,
            attributes: {},
            online: null,
            notifiable: null
        };
        this._initializationPromise = new Promise((resolve) => {
            this._resolveInitializationPromise = resolve;
        });
        if (configuration !== null) {
            this._resolveInitialization(configuration, identity, entityName, false);
        }
    }
    /**
     * User identity.
     */
    get identity() { return this.state.identity; }
    set identity(identity) { this.state.identity = identity; }
    set entityName(name) { this.state.entityName = name; }
    /**
     * Custom attributes of the user.
     */
    get attributes() { return this.state.attributes; }
    /**
     * Friendly name of the user, null if not set.
     */
    get friendlyName() { return this.state.friendlyName; }
    /**
     * Status of the real-time conversation connection of the user.
     */
    get online() { return this.state.online; }
    /**
     * User push notification registration status.
     */
    get notifiable() { return this.state.notifiable; }
    /**
     * True if this user is receiving real-time status updates.
     */
    get isSubscribed() { return this.subscribed == 'subscribed'; }
    // Handles service updates
    async _update(key, value) {
        await this._initializationPromise;
        let updateReasons = [];
        log.debug('User for', this.state.identity, 'updated:', key, value);
        switch (key) {
            case 'friendlyName':
                if (this.state.friendlyName !== value.value) {
                    updateReasons.push('friendlyName');
                    this.state.friendlyName = value.value;
                }
                break;
            case 'attributes':
                const updateAttributes = index.parseAttributes(value.value, `Retrieved malformed attributes from the server for user: ${this.state.identity}`, log);
                if (!isEqual__default['default'](this.state.attributes, updateAttributes)) {
                    this.state.attributes = updateAttributes;
                    updateReasons.push('attributes');
                }
                break;
            case 'reachability':
                if (this.state.online !== value.online) {
                    this.state.online = value.online;
                    updateReasons.push('online');
                }
                if (this.state.notifiable !== value.notifiable) {
                    this.state.notifiable = value.notifiable;
                    updateReasons.push('notifiable');
                }
                break;
            default:
                return;
        }
        if (updateReasons.length > 0) {
            this.emit('updated', { user: this, updateReasons: updateReasons });
        }
    }
    // Fetch reachability info
    async _updateReachabilityInfo(map, update) {
        await this._initializationPromise;
        if (!this.configuration.reachabilityEnabled) {
            return;
        }
        return map.get('reachability')
            .then(update)
            .catch(err => { log.warn('Failed to get reachability info for ', this.state.identity, err); });
    }
    // Fetch user
    async _fetch() {
        await this._initializationPromise;
        if (!this.state.entityName) {
            return this;
        }
        this.promiseToFetch = this.services.syncClient.map({ id: this.state.entityName, mode: 'open_existing', includeItems: true })
            .then(map => {
            this.entity = map;
            map.on('itemUpdated', args => {
                log.debug(`${this.state.entityName} (${this.state.identity}) itemUpdated: ${args.item.key}`);
                return this._update(args.item.key, args.item.data);
            });
            return Promise.all([
                map.get('friendlyName')
                    .then(item => this._update(item.key, item.data)),
                map.get('attributes')
                    .then(item => this._update(item.key, item.data)),
                this._updateReachabilityInfo(map, item => this._update(item.key, item.data))
            ]);
        })
            .then(() => {
            log.debug('Fetched for', this.identity);
            this.subscribed = 'subscribed';
            this.emit('userSubscribed', this);
            return this;
        })
            .catch(err => {
            this.promiseToFetch = null;
            throw err;
        });
        return this.promiseToFetch;
    }
    // Not private because it is accessed from Client constructor.
    async _ensureFetched() {
        await this._initializationPromise;
        return this.promiseToFetch || this._fetch();
    }
    /**
     * Edit user attributes.
     * @param attributes New attributes.
     */
    async updateAttributes(attributes) {
        await this._initializationPromise;
        if (this.subscribed == 'unsubscribed') {
            throw new Error('Can\'t modify unsubscribed object');
        }
        await this.services.commandExecutor.mutateResource('post', this.links.self, {
            attributes: JSON.stringify(attributes)
        });
        return this;
    }
    /**
     * Update the friendly name of the user.
     * @param friendlyName New friendly name.
     */
    async updateFriendlyName(friendlyName) {
        await this._initializationPromise;
        if (this.subscribed == 'unsubscribed') {
            throw new Error('Can\'t modify unsubscribed object');
        }
        await this.services.commandExecutor.mutateResource('post', this.links.self, {
            friendly_name: friendlyName
        });
        return this;
    }
    /**
     * Remove the user from the subscription list.
     * @return A promise of completion.
     */
    async unsubscribe() {
        await this._initializationPromise;
        if (this.promiseToFetch) {
            await this.promiseToFetch;
            this.entity.close();
            this.promiseToFetch = null;
            this.subscribed = 'unsubscribed';
            this.emit('userUnsubscribed', this);
        }
    }
    _resolveInitialization(configuration, identity, entityName, emitUpdated) {
        this.configuration = configuration;
        this.identity = identity;
        this.entityName = entityName;
        this.links = {
            self: `${this.configuration.links.users}/${this.identity}`
        };
        this._resolveInitializationPromise();
        if (emitUpdated) {
            this.emit('updated', {
                user: this,
                updateReasons: [
                    'friendlyName',
                    'attributes',
                    'online',
                    'notifiable'
                ]
            });
        }
    }
}
tslib_es6.__decorate([
    declarativeTypeValidator.validateTypesAsync(['string', 'number', 'boolean', 'object', declarativeTypeValidator.literal(null)]),
    tslib_es6.__metadata("design:type", Function),
    tslib_es6.__metadata("design:paramtypes", [Object]),
    tslib_es6.__metadata("design:returntype", Promise)
], User.prototype, "updateAttributes", null);
tslib_es6.__decorate([
    declarativeTypeValidator.validateTypesAsync('string'),
    tslib_es6.__metadata("design:type", Function),
    tslib_es6.__metadata("design:paramtypes", [Object]),
    tslib_es6.__metadata("design:returntype", Promise)
], User.prototype, "updateFriendlyName", null);

exports.User = User;
